synapse/types/__init__.py

```diff
@@ -797,16 +798,44 @@ def copy_and_advance(self, key: str, new_value: Any) -> "StreamToken":
             return new_token
 
         new_token = self.copy_and_replace(key, new_value)
-        new_id = int(getattr(new_token, key))
-        old_id = int(getattr(self, key))
+        new_id = new_token.get_field(key)
+        old_id = self.get_field(key)
 
         if old_id < new_id:
             return new_token
         else:
             return self
 
-    def copy_and_replace(self, key: str, new_value: Any) -> "StreamToken":
-        return attr.evolve(self, **{key: new_value})
+    def copy_and_replace(self, key: StreamKeyType, new_value: Any) -> "StreamToken":
+        return attr.evolve(self, **{key.value: new_value})
+
+    @overload
+    def get_field(self, key: Literal[StreamKeyType.ROOM]) -> RoomStreamToken:
+        ...
+
+    @overload
+    def get_field(
+        self,
+        key: Literal[
+            StreamKeyType.ACCOUNT_DATA,
+            StreamKeyType.DEVICE_LIST,
+            StreamKeyType.PRESENCE,
+            StreamKeyType.PUSH_RULES,
+            StreamKeyType.RECEIPT,
+            StreamKeyType.TO_DEVICE,
+            StreamKeyType.TYPING,
+            StreamKeyType.UN_PARTIAL_STATED_ROOMS,
+        ],
+    ) -> int:
+        ...
+
+    @overload
+    def get_field(self, key: StreamKeyType) -> Union[int, RoomStreamToken]:
+        ...
```

Comment:
I think this overload is unnecessary?

Reply:
(Since it matches the actual definition.)

Reply:
Annoyingly it's necessary to deal with callers where we don't put in a literal

Reply:
Ah, because the others are literals? That's frustrating it doesn't fallback to the real definition. Well anyway, seems fine to me. 👍 

(Aside: I wonder if `StreamKeyType` could be generic to the type of token it returns?)

Reply:
Yeah, annoyingly the "implementation" (i.e. the one that isn't marked as an overload) doesn't count, and you can't mark it as an overload.

Reply:
> (Aside: I wonder if `StreamKeyType` could be generic to the type of token it returns?)

I was literally just thinking on how we could do that sanely, but I'm failing to see how.

Reply:
I don't think you can do it reasonably, maybe with a type guard or something, but that seems quite annoying.